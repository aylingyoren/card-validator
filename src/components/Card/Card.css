.card-validator {
    padding: 20px;
    border: 1px solid rgba(50, 50, 50, .3);
    width: 340px;
}

.light-border {
    border: 1px solid rgba(250, 250, 250, .3);
}

.header, .dark-text {
    color:rgb(50, 50, 50);
}
  
.light-text {
    color:rgb(250, 250, 250);
}

.card-fields, .expDate-set, .cvv-set, .cvv-field, .month-field, .year-field, .expDate-cvv, .expDate-field {
    display: flex;
}
.card-fields, .expDate-set, .cvv-set, .cvv-field, .month-field, .year-field {
    flex-direction: column;
}

.card-fields, .expDate-set, .cvv-set {
    align-items: flex-start;
}

.card-field, .submit-btn {
    margin-bottom: 7px;
    width: 300px;
    height: 30px;
}

.card-field-label, .err-msg {
    font-size: 12px;
}

.card-field-label {
    margin-bottom: 5px;
    font-weight: 500;
}

.expDate-cvv {
    gap: 20px;
}

.expDate-input, .month-field .err-msg, .year-field .err-msg {
    width: 70px;
}

.expDate-field {
    gap: 7px;
}

.cvv-input {
    width: 100px;
}

.submit-btn {
    padding: 0;
    border: none;
    cursor: pointer;
}

.err-msg {
    margin-bottom: 20px;
    color: rgb(199, 12, 68);
}

.light-err {
    color: rgb(231, 102, 152);
}

.light-btn {
    background-color: rgb(142, 238, 230);
}