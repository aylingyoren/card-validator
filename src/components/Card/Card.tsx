import { Formik, Field, Form, ErrorMessage } from 'formik';
import * as Yup from "yup";
import CardView from '../CardView';
import "./Card.css";

export interface MyFormValues {
  cardNumber: string;
  cardHolder: string;
  CVV: string;
  month: string;
  year: string;
}

type CardProps = {
  isDarkModeActive: boolean
}

function Card({isDarkModeActive}: CardProps) {
  const initialValues: MyFormValues = { cardNumber: '', cardHolder: '', CVV: '', month: '', year: '' };

  return (
    <div>
        <h2 className={isDarkModeActive ? "header light-text" : "header"}>Visa Card</h2>
        
        <div className={isDarkModeActive ? "card-validator light-border" : "card-validator"}>
          <Formik
            initialValues={initialValues}
            validationSchema={Yup.object({
              cardNumber: Yup.string()
                .label('Card Number')
                .max(16)
                .min(16)
                .matches(/^[0-9]*$/, 'Only digits')
                .required('Required'),
              cardHolder: Yup.string()
                .label('Card Holder')
                .max(20, 'Must be 20 characters or less')
                .min(2, 'Must be 2 characters or more')
                .matches(/^[aA-zZ\s]+$/, 'Wrong format')
                .required('Required'),
              month: Yup.string()
                .label('Expiration Date')
                .required('Required'),
              year: Yup.string()
                .label('Expiration Date')
                .required('Required'),
              CVV: Yup.string()
                .label('CVV Number')
                .min(3, 'Must be 3 or 4 digits')
                .max(4, 'Must be 3 or 4 digits')
                .required('Required'),
            })}
            onSubmit={(values, {setSubmitting}) => {
              setTimeout(() => {
                alert(JSON.stringify(values, null, 2));
                setSubmitting(false);
              }, 500)
            }}
          >
            {({ values, isSubmitting }) => (
              <Form className='card-fields'>
                <CardView values={values} isDarkModeActive={isDarkModeActive} />
                <label className={isDarkModeActive ? 'card-field-label light-text' : 'card-field-label dark-text'} htmlFor="cardNumber">Card Number</label>
                <Field className='card-field' name="cardNumber" type="text" />
                <span className={isDarkModeActive ? 'err-msg light-err' : 'err-msg'}><ErrorMessage name="cardNumber" /></span>
              
                <label className={isDarkModeActive ? 'card-field-label light-text' : 'card-field-label dark-text'} htmlFor="cardHolder">Card Holder</label>
                <Field className='card-field' name="cardHolder" type="text" />
                <span className={isDarkModeActive ? 'err-msg light-err' : 'err-msg'}><ErrorMessage name="cardHolder" /></span>

                <div className="expDate-cvv">
                  <div className="expDate-set">
                    <label className={isDarkModeActive ? 'card-field-label light-text' : 'card-field-label dark-text'} htmlFor="month">Expiration Date</label>
                    <div className="expDate-field">
                      <div className="month-field">
                        <Field className='card-field expDate-input' as="select" name="month">
                          <option value="">Month</option>
                          <option value="01">January</option>
                          <option value="02">February</option>
                          <option value="03">March</option>
                          <option value="04">April</option>
                          <option value="05">May</option>
                          <option value="06">June</option>
                          <option value="07">July</option>
                          <option value="08">August</option>
                          <option value="09">September</option>
                          <option value="10">October</option>
                          <option value="11">November</option>
                          <option value="12">December</option>
                        </Field>
                        <span className={isDarkModeActive ? 'err-msg light-err' : 'err-msg'}><ErrorMessage name="month" /></span>
                      </div>

                      <div className="year-field">
                        <Field className='card-field expDate-input' as="select" name="year">
                          <option value="">Year</option>
                          <option value="27">2027</option>
                          <option value="26">2026</option>
                          <option value="25">2025</option>
                          <option value="24">2024</option>
                          <option value="23">2023</option>
                          <option value="22">2022</option>
                        </Field>
                        <span className={isDarkModeActive ? 'err-msg light-err' : 'err-msg'}><ErrorMessage name="year" /></span>
                      </div>
                    </div>
                    
                  </div>
                
                  <div className="cvv-set">
                    <label className={isDarkModeActive ? 'card-field-label light-text' : 'card-field-label dark-text'} htmlFor="CVV">CVV</label>
                    <Field className='card-field cvv-input' name="CVV" type="text" />
                    <span className={isDarkModeActive ? 'err-msg light-err' : 'err-msg'}><ErrorMessage name="CVV" /></span>
                  </div>
                </div>
              
                <button className={isDarkModeActive ? "submit-btn light-btn": "submit-btn dark"} type="submit" disabled={isSubmitting}>Submit</button>
              </Form>
            )}
          </Formik>
        </div>
    </div>
  );
}

export default Card;