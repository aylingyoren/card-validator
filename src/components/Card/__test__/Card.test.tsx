import {
  render,
  screen,
  fireEvent,
  waitFor,
  getByTestId,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import Card, { MyFormValues, CardProps } from "../Card";

function renderCard(props: Partial<CardProps> = {}) {
  const defaultProps: CardProps = {
    isDarkModeActive: false,
  };
  return render(<Card {...defaultProps} {...props} />);
}

test("should display a blank card form, with dark theme set by default", async () => {
  const { findByTestId } = renderCard();
  const container = await findByTestId("card-fields");

  const submitBtn = await findByTestId("submit-btn");
  const header = await findByTestId("header");
  const input = getByTestId(container, "cardNumberInput");

  input.focus();
  expect(input).toHaveFocus();
  //   screen.getByLabelText("Card Number"); //doesn't find it although it's in the dom
  screen.getByText("Submit");

  expect(submitBtn).not.toHaveAttribute("disabled"); // disabled === 'null' ????
  expect(header).toHaveClass("header");
  expect(header).not.toHaveClass("header light-text");

  expect(container).toHaveFormValues({
    cardNumber: "",
    cardHolder: "",
    month: "",
    year: "",
    CVV: "",
  });

  render(<Card isDarkModeActive />);
});

it("renders correctly when there are no items", () => {
  const tree = renderer.create(<Card isDarkModeActive />).toJSON();
  expect(tree).toMatchSnapshot();
});
