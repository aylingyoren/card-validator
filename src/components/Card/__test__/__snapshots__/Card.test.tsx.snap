// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div>
  <h2
    className="header light-text"
    data-testid="header"
  >
    Visa Card
  </h2>
  <div
    className="card-validator light-border"
  >
    <form
      action="#"
      className="card-fields"
      data-testid="card-fields"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="card-body light-card-border"
        data-testid="card-body"
      >
        <h2
          className="card-number"
          data-testid="card-number"
        >
          0000 0000 0000 0000
        </h2>
        <div
          className="card-holder"
          data-testid="card-holder"
        >
          NAME SURNAME
        </div>
        <div
          className="exp-date"
          data-testid="exp-date"
        >
          00
           / 
          00
        </div>
        <div
          className=""
          data-testid="cvv"
        >
          CVV: 
          000(0)
        </div>
        <img
          alt="card chip"
          className=""
          data-testid="chip-img"
          src="chip-card.png"
        />
        <img
          alt="visa logo"
          className=""
          data-testid="visa-img"
          src="visa.png"
        />
      </div>
      <label
        aria-labelledby="cardNumber"
        className="card-field-label light-text"
        data-testid="cardNumber"
        htmlFor="cardNumber"
      >
        Card Number
      </label>
      <input
        className="card-field"
        data-testid="cardNumberInput"
        name="cardNumber"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
      <span
        className="err-msg light-err"
      />
      <label
        className="card-field-label light-text"
        htmlFor="cardHolder"
      >
        Card Holder
      </label>
      <input
        className="card-field"
        name="cardHolder"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
      <span
        className="err-msg light-err"
      />
      <div
        className="expDate-cvv"
      >
        <div
          className="expDate-set"
        >
          <label
            className="card-field-label light-text"
            htmlFor="month"
          >
            Expiration Date
          </label>
          <div
            className="expDate-field"
          >
            <div
              className="month-field"
            >
              <select
                className="card-field expDate-input"
                name="month"
                onBlur={[Function]}
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  Month
                </option>
                <option
                  value="01"
                >
                  January
                </option>
                <option
                  value="02"
                >
                  February
                </option>
                <option
                  value="03"
                >
                  March
                </option>
                <option
                  value="04"
                >
                  April
                </option>
                <option
                  value="05"
                >
                  May
                </option>
                <option
                  value="06"
                >
                  June
                </option>
                <option
                  value="07"
                >
                  July
                </option>
                <option
                  value="08"
                >
                  August
                </option>
                <option
                  value="09"
                >
                  September
                </option>
                <option
                  value="10"
                >
                  October
                </option>
                <option
                  value="11"
                >
                  November
                </option>
                <option
                  value="12"
                >
                  December
                </option>
              </select>
              <span
                className="err-msg light-err"
              />
            </div>
            <div
              className="year-field"
            >
              <select
                className="card-field expDate-input"
                name="year"
                onBlur={[Function]}
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  Year
                </option>
                <option
                  value="27"
                >
                  2027
                </option>
                <option
                  value="26"
                >
                  2026
                </option>
                <option
                  value="25"
                >
                  2025
                </option>
                <option
                  value="24"
                >
                  2024
                </option>
                <option
                  value="23"
                >
                  2023
                </option>
                <option
                  value="22"
                >
                  2022
                </option>
              </select>
              <span
                className="err-msg light-err"
              />
            </div>
          </div>
        </div>
        <div
          className="cvv-set"
        >
          <label
            className="card-field-label light-text"
            htmlFor="CVV"
          >
            CVV
          </label>
          <input
            className="card-field cvv-input"
            name="CVV"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
          <span
            className="err-msg light-err"
          />
        </div>
      </div>
      <button
        className="submit-btn light-btn"
        data-testid="submit-btn"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
